WIP: Generate protobuf IDL from Go versioned structs

smarterclayton opened the pull request:
First stage of work to allow our versioned Go structs to convert to protobuf IDL, which will be followed by generating the IDL back into serializers/deserializers for the native Go structs from protobuf.  With that, we can support multiple codecs from the master behind an Accept / Content-Type replacement.

Assumptions:
- We want our Go structs to be canonical (minimal impact to existing code)
- We want to expose protobuf serialization for memory / CPU benefits
- Only our external types need to be convertible to protobuf

Proposed path:
- Use go2idl to generate a set of `.proto` IDLs (one for each API package) as `pkg/api/v1/generated.proto`
  - Have `pkg/api/unversioned` be the catch all for types like Time, etc
  - Ensure that when you generate an IDL for a package, you can identify import packages (and we import the types into the IDL rather than redefining them)
  - Respect an existing `protobuf` tag on the Go struct so we get tag value stability
  - Possibly give each package a shortname (although we could preserve the full path), like `k8s.io/kubernetes/pkg/api/v1` -> `k8s.api.v1`
- Use protoc + gogoprotobuf + extensions to generate a set of Go structs and generated marshallers/unmarshallers into `pkg/api/v1/generated.pb.go`
  - Custom marshallers have to be attached to their type
  - Verify that the Go structs are _exactly_ like the source structs (round trip to IDL should be exact) with the addition of the `protobuf` struct tags
    - If not, we may need to make changes to gogo protobuf
- Copy the protobuf struct tags back onto the source objects (so that proto tag order is preserved)
- Remove the generated structs
- Test and verify that the generated marshallers work against our types and can round trip perfectly.
- Implement custom marshallers for things outside of our package and embed them in `unversioned`

Rough prototype path:
- [x] generate an IDL from the go struct (leveraging existing protobuf tag or json tag, generate otherwise)
- [ ] ensure we can scope generation down to only the types defined in the field
  - [ ] types that are outside the Go package but inside the current package have the wrong proto package
  - [ ] skip fundamental types cleanly
- [ ] sweep all fields to ensure we can generate the types, identify types that need custom serializations
  - [ ] util.Time
  - [ ] handle enums
- [ ] prove we can generate an identical struct
  - [ ] investigate which gogoprotobuf features we need to ensure the generate Go struct would match the input go struct
    - [X] nullable - able to generate embedded structs correctly
    - [X] casttype - able to generate alias field names correctly and force a cast
- [ ] if types are equivalent, merge protobuf fields back into canonical structs
- [ ] investigate generating marshal / unmarshal for protobuf without generating struct (separate file?)
- [ ] set up round trip fuzz tester and verify Go struct -> protobuf -> Go struct is lossless
- [ ] copy comments

smarterclayton committed:
Abstract file generation to allow alternate structures

A Protobuf IDL has similar structure, but different output, than a
golang file. Formalize that behind a FileType interface.

smarterclayton committed:
WIP: go struct to protobuf IDL with reentrancy

Use the existing structs to generate a protobuf IDL from a set of Go
structs. Allow gogoprotobuf protobuf struct tags to define default
values for the generation (which will allow eventual roundtripping of
struct -> IDL -> gogoprotobuf generation), and use JSON tags as a
fallback.

mikedanese commented:
#8132

lavalamp commented:
Oh wow, I am super excited about this! Will review in detail later.

smarterclayton commented:
It's still pretty rough, but a) it gave me a chance to learn go2idl and make sure I understood the ramifications of the namer subsystem (i'm not sure I applied it as intended, yet), and b) at least verify we can generate approximately equal structs from gogoprotobuf.

lavalamp commented on cmd/libs/go2idl/go-to-protobuf/generator.go:
```diff
@@ -0,0 +1,488 @@
+func (g *genProtoIDL) Namers(c *generator.Context) namer.NameSystems {
+	imports := generator.NewImportTracker()
+	// Have the raw namer for this file track what it imports.
+	return namer.NameSystems{"raw": namer.NewRawNamer(imports)}
```
If you don't need to track imports, just pass nil.

lavalamp commented on cmd/libs/go2idl/go-to-protobuf/generator.go:
```diff
@@ -0,0 +1,488 @@
+func (g *genProtoIDL) Filename() string { return g.OptionalName + ".proto" }
+func (g *genProtoIDL) FileType() string { return "protoidl" }
+func (g *genProtoIDL) Namers(c *generator.Context) namer.NameSystems {
+	imports := generator.NewImportTracker()
```
It only makes sense to make this if you use it later :)

lavalamp commented on cmd/libs/go2idl/go-to-protobuf/generator.go:
```diff
@@ -0,0 +1,488 @@
+	}
+	fields, err := membersToFields(b.t, b.localPackage)
+	if err != nil {
+		sw.Do(fmt.Sprintf("// ERROR: type $.Name$ cannot be converted to protobuf: %v\n", err), b.t)
```
`sw.Do("// ERROR: type $type.Name$ cannot be converted to protobuf: $.error$\n", err), map[string]interface{}{"type": b.t, "error": err.Error()})`

It's possible to construct a helper function so that the last bit can be condensed to e.g. `args("type", b.t, "error", err.Error())`

smarterclayton replied:
Is it possible to inject non-namer functions into the template?  Didn't
seem obvious how if so.

On Nov 2, 2015, at 4:00 PM, Daniel Smith <notifications@github.com> wrote:

In cmd/libs/go2idl/go-to-protobuf/generator.go
<https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43681338>:

> +		field.Tag = highest
> +		byTag[field.Tag] = field
> +	}
> +	return fields, nil
> +}
> +
> +func (b bodyGen) doStruct(sw *generator.SnippetWriter) {
> +	if len(b.t.Name.Name) == 0 {
> +		return
> +	}
> +	if isPrivateGoName(b.t.Name.Name) {
> +		return
> +	}
> +	fields, err := membersToFields(b.t, b.localPackage)
> +	if err != nil {
> +		sw.Do(fmt.Sprintf("// ERROR: type $.Name$ cannot be converted to protobuf: %v\n", err), b.t)

sw.Do("// ERROR: type $type.Name$ cannot be converted to protobuf:
$.error$\n", err), map[string]interface{}{"type": b.t, "error":
err.Error()})

It's possible to construct a helper function so that the last bit can be
condensed to e.g. args("type", b.t, "error", err.Error())

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16649/files#r43681338>.

lavalamp replied:
You'd have to modify the SnippetWriter, but it would certainly be possible.

lavalamp commented on cmd/libs/go2idl/go-to-protobuf/main.go:
```diff
@@ -0,0 +1,92 @@
+var (
+	inputPackages = flag.StringP("input-packages", "i" /*"k8s.io/kubernetes/pkg/api,k8s.io/kubernetes/pkg/apis/extensions"*/, "", "comma-separated list of directories to get input types from.")
+	packages      = flag.StringP("packages", "p", "k8s.io/kubernetes/pkg/api/unversioned=k8s.api.unversioned,k8s.io/kubernetes/pkg/api/v1=k8s.api.v1,k8s.io/kubernetes/pkg/apis/extensions/v1beta1=k8s.api.extensions.v1beta1", "comma-separated list of directories to get input types from.")
+	outputBase    = flag.StringP("output-base", "o", filepath.Join(os.Getenv("GOPATH"), "src"), "Output base; defaults to $GOPATH/src/")
```
As @wojtek-t said elsewhere, we should really extract these out to a common place. I'll try to add that.

lavalamp replied:
In #16637.

lavalamp commented on cmd/libs/go2idl/types/types.go:
```diff
@@ -275,6 +275,18 @@ var (
 		Name: Name{Name: "uint"},
 		Kind: Builtin,
 	}
+	Float64 = &Type{
```
Wow, I totally forgot about floats! We can keep pretending complex types don't exist ;)

smarterclayton replied:
My thought exactly :)

On Nov 2, 2015, at 4:05 PM, Daniel Smith <notifications@github.com> wrote:

In cmd/libs/go2idl/types/types.go
<https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43681965>:

> @@ -275,6 +275,18 @@ var (
>  		Name: Name{Name: "uint"},
>  		Kind: Builtin,
>  	}
> +	Float64 = &Type{

Wow, I totally forgot about floats! We can keep pretending complex types
don't exist ;)

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16649/files#r43681965>.

lavalamp commented on cmd/libs/go2idl/go-to-protobuf/package.go:
```diff
@@ -0,0 +1,155 @@
+func newProtobufPackage(packagePath, packageName string) *protobufPackage {
+	pkg := &protobufPackage{
+		PackageName: packageName,
+		PackagePath: packagePath,
```
Maybe put generated files in a different directory? (.../generated, .../proto, etc)

smarterclayton replied:
For the proto idl it felt right to have it in the same go package, but I'm
not locked into it.  It also means the proto import path is consistent with
the package name of the API (k8s.api.unversioned) vs having a different
import path than the package.

When we have the generated marshallers, those *have* to be in the same
package (since MarshalTo is a method on the struct).  The alternatives are
wrappers we add with a type switch just prior to serialization, but that is
somewhat yucky too.  But not there yet so can defer it till then.

On Nov 2, 2015, at 4:10 PM, Daniel Smith <notifications@github.com> wrote:

In cmd/libs/go2idl/go-to-protobuf/package.go
<https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43682658>:

> +
> +import (
> +	"fmt"
> +	"os"
> +	"path/filepath"
> +	"sort"
> +	"strings"
> +
> +	"k8s.io/kubernetes/cmd/libs/go2idl/generator"
> +	"k8s.io/kubernetes/cmd/libs/go2idl/types"
> +)
> +
> +func newProtobufPackage(packagePath, packageName string) *protobufPackage {
> +	pkg := &protobufPackage{
> +		PackageName: packageName,
> +		PackagePath: packagePath,

Maybe put generated files in a different directory? (.../generated,
.../proto, etc)

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16649/files#r43682658>.

lavalamp replied:
I see. I don't feel too strongly about the proto files. We need to be careful with the generated functions, like maybe deleting generated files before running the generator. We don't want to get into today's situation where we can't run the generator because it doesn't compile.

smarterclayton replied:
Agreed.  Proto is easier since it won't change the compilation state, but
the subsequent generation should purge before parse at a minimum.  We can
also look into generating a codec outside of the package.

On Nov 2, 2015, at 7:59 PM, Daniel Smith <notifications@github.com> wrote:

In cmd/libs/go2idl/go-to-protobuf/package.go
<https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43705760>:

> +
> +import (
> +	"fmt"
> +	"os"
> +	"path/filepath"
> +	"sort"
> +	"strings"
> +
> +	"k8s.io/kubernetes/cmd/libs/go2idl/generator"
> +	"k8s.io/kubernetes/cmd/libs/go2idl/types"
> +)
> +
> +func newProtobufPackage(packagePath, packageName string) *protobufPackage {
> +	pkg := &protobufPackage{
> +		PackageName: packageName,
> +		PackagePath: packagePath,

I see. I don't feel too strongly about the proto files. We need to be
careful with the generated functions, like maybe deleting generated files
before running the generator. We don't want to get into today's situation
where we can't run the generator because it doesn't compile.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16649/files#r43705760>.

lavalamp replied:
Making codecs outside of the package would be my first choice, but maybe
not if it's more than 20% extra effort.

On Mon, Nov 2, 2015 at 10:01 PM, Clayton Coleman <notifications@github.com>
wrote:

> In cmd/libs/go2idl/go-to-protobuf/package.go
> <https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43718532>
> :
>
> > +
> > +import (
> > +	"fmt"
> > +	"os"
> > +	"path/filepath"
> > +	"sort"
> > +	"strings"
> > +
> > +	"k8s.io/kubernetes/cmd/libs/go2idl/generator"
> > +	"k8s.io/kubernetes/cmd/libs/go2idl/types"
> > +)
> > +
> > +func newProtobufPackage(packagePath, packageName string) *protobufPackage {
> > +	pkg := &protobufPackage{
> > +		PackageName: packageName,
> > +		PackagePath: packagePath,
>
> Agreed. Proto is easier since it won't change the compilation state, but
> the subsequent generation should purge before parse at a minimum. We can
> also look into generating a codec outside of the package. On Nov 2, 2015,
> at 7:59 PM, Daniel Smith <notifications@github.com> wrote: In
> cmd/libs/go2idl/go-to-protobuf/package.go <#16649 (comment)
> <https://github.com/kubernetes/kubernetes/pull/16649#discussion_r43705760>
> >:
> + +import ( + "fmt" + "os" + "path/filepath" + "sort" + "strings" + + "
> k8s.io/kubernetes/cmd/libs/go2idl/generator" + "
> k8s.io/kubernetes/cmd/libs/go2idl/types" +) + +func
> newProtobufPackage(packagePath, packageName string) *protobufPackage { +
> pkg := &protobufPackage{ + PackageName: packageName, + PackagePath:
> packagePath,
> I see. I don't feel too strongly about the proto files. We need to be
> careful with the generated functions, like maybe deleting generated files
> before running the generator. We don't want to get into today's situation
> where we can't run the generator because it doesn't compile. — Reply to
> this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/16649/files#r43705760>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16649/files#r43718532>.
>

lavalamp commented:
So far looks quite promising!

smarterclayton committed:
Allow Path info to be carried with types.Name

Packages may have a path, a name, and a local package in proto

smarterclayton committed:
Remove tabs for output

smarterclayton committed:
Time changes

smarterclayton committed:
godep(gogo/protobuf/sortkeys):added

smarterclayton committed:
Back to proto2

smarterclayton committed:
Update conversions to handle primitive conversion

smarterclayton committed:
Unversioned types should not use ambiguous go-int

All external types that are not int64 are now marked as int32,
including
IntOrString. Prober is now int32 (43 years should be enough of an initial
probe time for anyone).

Did not change the metadata fields for now.

smarterclayton committed:
Generated code

smarterclayton committed:
Fix tests to pass with changed external types

smarterclayton committed:
Generated docs

smarterclayton committed:
Review comments

feiskyer commented:
Excited about this effort.  #17048  c/s runtime can use this to generate IDLs.

smarterclayton committed:
Tag RawExtension as being explicitly Protogenned

smarterclayton committed:
First stab at codec

timothysc commented:
/cc @wojtek-t

smarterclayton commented:
A few snags I'm working through now:

* had to convert all external types to fixed width ints (so that protobuf is correct), also updated api-conventions to make that a requirement for reviewers
* gogo/protobuf chokes on maps with type aliases, keys that are aliases, and embedded values (gogo/protobuf#116, gogo/protobuf#117, gogo/protobuf#118).  Looking at what I can do to fix that.
* Considering whether to omit TypeMeta from the inside of the protobuf objects, and instead wrap it on the outside.

smarterclayton committed:
More command

smarterclayton committed:
resource.Quantity protobuf

smarterclayton committed:
Codec

smarterclayton committed:
Serialization

smarterclayton committed:
Generated

smarterclayton committed:
Cmd

smarterclayton committed:
Updated generated

smarterclayton commented:
Did some preliminary benchmarks against standard fuzzed Pod with the protobuf generated serializers.  Output json is ~2k, output Proto is 1k.  

Net result - protobuf crushes the competition by almost an order of magnitude in speed and allocations.

```
BenchmarkEncodeCodec-8      	   50000	     31164 ns/op	    4969 B/op	      58 allocs/op # codec overhead, json, no conversion
BenchmarkEncodeJSONMarshal-8	   50000	     30241 ns/op	    7887 B/op	      49 allocs/op # golang json marshal
BenchmarkEncodeProtobuf-8   	  500000	      3823 ns/op	    1196 B/op	       6 allocs/op # protobuf marshal
```

Encode ~9x faster, ~9x less allocations, 7k less garbage

````
BenchmarkDecodeCodec-8           	   10000	    101062 ns/op	   17446 B/op	     352 allocs/op # includes conversion
BenchmarkDecodeIntoCodec-8       	   10000	    249900 ns/op	   32738 B/op	    1044 allocs/op # includes conversion
BenchmarkDecodeIntoJSON-8        	   20000	     86253 ns/op	    7483 B/op	     126 allocs/op # Golang JSON decoding, no conversion
BenchmarkDecodeIntoJSONCodecGen-8	   30000	     48856 ns/op	    9678 B/op	     240 allocs/op # ugorji decoding, no conversion
BenchmarkDecodeIntoProtobuf-8    	  200000	      6097 ns/op	    3560 B/op	      64 allocs/op # proto decoding, no conversion
```

Decode 6x faster than ugorji, 4x less allocations.

This is before some fairly no brainer Proto optimizations.  Using protobuf enums instead of strings on many of our default API objects could result in a smaller size AND less allocations, since []byte -> string is one allocation - that might cut 10-20 allocations off a complex non-fuzzed object (fuzzed objects have random enum fields, end user objects would have only valid fields).

I suspect that with this change we would get down to the point that codec inefficiencies dominate the results (yay!) - we are still doing a few reflect allocations in conversion and decode that can be avoided (invoking Convert(interface{} interface{}) requires a reflection allocation, we should invoke static methods with a map lookup or something).

Once I get the gogo/protobuf upstream fixes in for maps that have aliased keys, and sort out a few decode bugs I've hit (probably caused by the hacks necessary to work around aliased maps), i'll detail the next steps.

smarterclayton committed:
WIP - hacks to get to a runnable test state

smarterclayton commented:
Here's what the v1 IDL looks like: https://github.com/kubernetes/kubernetes/blob/41d9c53ed253848f9160574b795d8bf66e9615e4/pkg/api/v1/generated.proto - go2idl made it easy to carry over comments (yay @lavalamp's code)

smarterclayton commented:
One downside - the marshalers add about 60k LoC.  Adds another 20-40% or so to our binary size

smarterclayton commented:
Concept has been proved, moving to more formal process in #17854
